okhttp/src/main/java/okhttp3/internal/concurrent/TaskQueue.kt

```diff
@@ -39,12 +62,126 @@ interface TaskQueue {
    * is running when that time is reached, that task is allowed to complete before this task is
    * started. Similarly the task will be delayed if the host lacks compute resources.
    */
-  fun schedule(task: Task, delayNanos: Long = 0L)
+  fun schedule(task: Task, delayNanos: Long) {
+    task.initQueue(this)
+
+    synchronized(taskRunner) {
+      if (scheduleAndDecide(task, delayNanos)) {
+        taskRunner.kickCoordinator(this)
+      }
+    }
+  }
+
+  /** Adds [task] to run in [delayNanos]. Returns true if the coordinator should run. */
+  private fun scheduleAndDecide(task: Task, delayNanos: Long): Boolean {
+    val now = taskRunner.backend.nanoTime()
+    val executeNanoTime = now + delayNanos
+
+    // If the task is already scheduled, take the earlier of the two times.
+    val existingIndex = futureTasks.indexOf(task)
+    if (existingIndex != -1) {
+      if (task.nextExecuteNanoTime <= executeNanoTime) return false // Already scheduled earlier.
+      futureTasks.removeAt(existingIndex) // Already scheduled later: reschedule below!
+    }
+    task.nextExecuteNanoTime = executeNanoTime
+
+    // Insert in chronological order.
+    var insertAt = futureTasks.indexOfFirst { it.nextExecuteNanoTime - now > delayNanos }
```

Comment:
nit: simpler to compare to task.nextExecuteNanoTime? without backing out now?

Reply:
Unfortunately `System.nanoTime()` is documented to permit wrapping around. 

For example, `now` could be `Long.MAX_VALUE - 100`, the existing task could be `Long.MAX_VALUE` and the task-being scheduled could be `Long.MAX_VALUE + 100` which is negative. We want to subtract off now to compare `100 < 200` instead of comparing the values-as-is which yields a different result.

I’ll add a comment!

Reply:
Watch out for

"Differences in successive calls that span greater than approximately 292 years (263 nanoseconds) will not correctly compute elapsed time due to numerical overflow. "

:)

Reply:
I’m not too worried about scheduling things 292 years into the future. I am worried about what the nanoTime initial value is. Is it 0?

Reply:
Yep, wasn't arguing against your valid justification.  Just going back to basics and re-reading nanoTime, where I found that gem of a warning.